### PR TITLE
Sm/bundle-delete-nuts

### PR DESCRIPTION
**Pull request by mshanemc**
_draft as it requires https://github.com/forcedotcom/source-tracking/pull/55_

### What does this PR do?
fixes json output formatting for push commands (didn't match toolbelt)
new option in sfdx-project.json for pushPackageDirectoriesSequentially
use new STL logic for byPkgDir
iterate array of componentSets for push
NUT for asserting how many deploys happened from the org
NUTs for LWC bundle-member-deletion @W-10216279@

What issues does this PR fix or reference?
@W-10146239@ 

---

QA notes:

top-level boolean prop pushPackageDirectoriesSequentially on sfdx-project.json

use this updated push command to QA the STL changes
I used https://github.com/salesforcecli/sample-project-multiple-packages as a test repo

check what happens with duplicate files in json output from subsequent deployments--do you get duplicates? Is that different from toolbelt or the same?

testing ideas that I didn't try
examining hook output (should fire once for each directory)
check source tracking files (source:status) on an mpd deployment
check pushPackageDirectoriesSequentially when there's only one pkgDirectory
test the human (non-json) output with a MPD project...the messages and fileResponses should all merge into one table across all the deploys
have a failure on any one pkgDir...it should stop any other deployments from happening.

### CODE REVIEW

File: test/nuts/trackingCommands/mpd-sequential.nut.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { fs, AuthInfo, Connection } from '@salesforce/core';
+import { expect } from 'chai';
+import { TestSession, execCmd } from '@salesforce/cli-plugins-testkit';
+import { replaceRenamedCommands } from '@salesforce/source-tracking';
+import { PushResponse } from '../../../src/formatters/pushResultFormatter';
+
+let session: TestSession;
+let conn: Connection;
+
+describe('end-to-end-test for tracking with an org (single packageDir)', () => {
```
Comment (WillieRuemmele) [inline test/nuts/trackingCommands/mpd-sequential.nut.ts:20]: this is using an MPD project, not `(single packageDir)` right?

do we need a NUT testing `"pushPackageDirectoriesSequentially": false`? Are we validating that `"true"` and `true` result in the same behavior?

**Review by WillieRuemmele — APPROVED** [review]
nothing blocking - just comments/questions

File: src/commands/force/source/beta/push.ts [hunk 2]
```diff
@@ -76,57 +78,65 @@ export default class Push extends DeployCommand {
...
-      this.logger.warn('There are no changes to deploy');
-      return;
-    }
-
-    // fire predeploy event for sync and async deploys
-    await this.lifecycle.emit('predeploy', componentSet.toArray());
+  }
 
+  protected async deploy(): Promise<void> {
+    const isMPD = getBoolean(await this.project.resolveProjectConfig(), 'pushPackageDirectoriesSequentially');
```
[the remark this answers is not in the document]
Comment (mshanemc) [reply src/commands/force/source/beta/push.ts:84]: default false

File: test/formatters/pushResultFormater.test.ts [hunk 3]
```diff
@@ -59,20 +63,30 @@ describe('PushResultFormatter', () => {
 
   describe('human output', () => {
     it('returns expected output for success', () => {
+      process.exitCode = 0;
```
[the remark this answers is not in the document]
Comment (mshanemc) [reply test/formatters/pushResultFormater.test.ts:66]: They're set in the command (resolveSuccess method).  The formatter for push looks at the exit code so for its UT it needs to be set.

An alternative would be to pass the exitCode to the formatter from the command.

File: test/nuts/trackingCommands/mpd-sequential.nut.ts (hunk 1 above)
Comment (mshanemc) [reply test/nuts/trackingCommands/mpd-sequential.nut.ts:20]: 1. it's *the* mpd project :)
2. having a non-sequential mpd is a good idea.
3. "true" would result in it not being mpd.  It's really a boolean, not a boolean | string.  Maybe that's worth a discussion...

**Pull request merged**